File: problems/0003_longest-substring-without-repeating-characters/map/Solution.py
#!/usr/bin/python3

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d = {}
        m, start, end = 0, 0, 0
        for c in s:
            idx = d.get(c)
            d[c] = end
            end += 1
            if idx is None or idx < start:
                m = max(m, end - start)
            else:
                start = idx + 1
        return m

File: problems/0003_longest-substring-without-repeating-characters/map/solution.c
int lengthOfLongestSubstring(char * s) {
    int map[] = { [0 ... 255] = -1 };
    int max = 0, start = 0, end = 0;
    for (char c = *s; c != '\0'; c = *(s + end)) {
        const int idx = map[c];
        map[c] = end++;
        if (idx == -1 || idx < start) {
            const int l = end - start;
            if (l > max)
                max = l;
        } else
            start = idx + 1;
    }
    return max;
}
